enhancements/network/on-prem-mutable-vips.md

```diff
@@ -0,0 +1,199 @@
+
+1. After keepalived configuration is changed, the service is restarted or reloaded to apply the changes.
+
+### API Extensions
+
+New fields will have to be added to the platform spec section for [baremetal](https://github.com/openshift/api/blob/938af62eda38e539488d6193e7af292e47a09a5e/config/v1/types_infrastructure.go#L694), [vsphere](https://github.com/openshift/api/blob/938af62eda38e539488d6193e7af292e47a09a5e/config/v1/types_infrastructure.go#L1089) and [OpenStack](https://github.com/openshift/api/blob/938af62eda38e539488d6193e7af292e47a09a5e/config/v1/types_infrastructure.go#L772). In the first implementation only baremetal will be covered.
+
+### Implementation Details/Notes/Constraints
+
+Because of the expertise and the team owning this feature, we think the best place for implementing the logic is Cluster Network Operator. CNO however is not an owner of the `cluster` Infrastructure CR what at first sight may seem like a challenge.
+
+Kubernetes API provide a solution for this and we can leverage Watches (with Predicate for optimization) to allow a controller inside CNO to watch an object it does not own. This in fact already happens inside CNO in the [infrastructureconfig controller](https://github.com/openshift/cluster-network-operator/blob/2005bcd8c93de5bffc05c9c943b51386007f6b9a/pkg/controller/infrastructureconfig/infrastructureconfig_controller.go#L47).
+
+The controller would get a logic implemented that validates if the change requested by user is valid as in the first implementation we only want to allow adding second pair of VIPs but forbid any modifications of the already configured ones.
+
+The current values of `spec.platformStatus.[*].xxx` are propagated during installation by XXX. After implementing this feature, the same XXX would also set `spec.platformSpec.[*].xxx` so that the cluster is installed with a correct configuration.
+
+When modifying `spec.platformSpec.[*].xxx` the CNO controller would need to propagate this change down to the [keepalived templates managed by the Machine Config Operator](https://github.com/openshift/machine-config-operator/blob/2b31b5b58f0d7d9fe6c3e331e4b3f01c9a1bd00c/templates/common/on-prem/files/keepalived.yaml#L46-L49).
+
+### Risks and Mitigations
+
+Minimal risk. The dual-stack VIPs feature is already used. This is just adding an ability to add this feature to an already existing cluster. Because modifying and deleting VIPs is out of scope of this enhancement, the operations with the biggest potential of causing an issue do not need to be covered.
+
+### Drawbacks
+
+Because we do not implement Update and Delete operations now, the main drawback is a scenario when user adds an incorect address to be configured. We will not allow to modify the entry, therefore enablement of this feature needs to be performed with some level of care by the end user.
```

Comment:
Hmm, good point. We don't currently allow updating this if a mistake is made during initial deployment either, but if you deploy a cluster and immediately realize you picked the wrong VIP it's much less painful to re-deploy. That may not be an option for a long-lived cluster.

I wonder how hard to it would be to allow deleting only the second VIP. I think that's still a safe change from the keepalived perspective, and you could do a delete and create if you made a mistake in this process.

Reply:
Updated so that we will allow Create and Delete